pkg/solana/txm/txm.go

```diff
@@ -303,65 +295,224 @@ func (txm *Txm) sendWithRetry(ctx context.Context, msg pendingTx) (solanaGo.Tran
+	getCurrentHeight := func(ctx context.Context) (uint64, error) {
+		client, errClient := txm.client.Get()
+		if errClient != nil {
+			return 0, errClient
+		}
+
+		currHeight, errSlot := client.SlotHeight(ctx)
+		if errSlot != nil {
+			return 0, errSlot
+		}
+
+		return currHeight, nil
+	}
+
+	newBumpTime = bumpTime
+	currHeight, err := getCurrentHeight(ctx)
+	if err != nil {
+		return false, false, bumpTime, fmt.Errorf("failed to get current height: %w", err)
+	}
+
+	// Check if fee bumping is enabled and decide based on the criteria
+	if txm.cfg.FeeBumpPeriod() > 0 {
+		switch txm.cfg.FeeBumpCriteria() {
+		case FeeBumpCriteriaFixedIntervals:
+			// Bump the fee at fixed intervals. We may want to bump without blockhash being expired.
+			if time.Since(bumpTime) >= txm.cfg.FeeBumpPeriod() {
+				shouldBump = true
+				newBumpTime = time.Now()
 			}
 
-			// exponential increase in wait time, capped at 250ms
-			deltaT *= 2
-			if deltaT > MaxRetryTimeMs {
-				deltaT = MaxRetryTimeMs
+			// in case blockhash has expired, we need to update it.
+			if txm.txs.IsBlockhashExpired(txID, currHeight) {
+				needBlockhashUpdate = true
 			}
-			tick = time.After(time.Duration(deltaT) * time.Millisecond)
+		case FeeBumpCriteriaExpiration:
+			// Bump the fee only if the blockhash has expired
+			if txm.txs.IsBlockhashExpired(txID, currHeight) {
+				shouldBump = true
+				needBlockhashUpdate = true
+			}
+		case FeeBumpCriteriaNone:
+			// Do not bump the fee
+			shouldBump = false
+		default:
+			txm.lggr.Errorw("unknown fee bump criteria", "criteria", txm.cfg.FeeBumpCriteria())
 		}
-	}(ctx, baseTx, initTx)
+	}
 
-	// return signed tx, id, signature for use in simulation
-	return initTx, msg.id, sig, nil
+	return shouldBump, needBlockhashUpdate, newBumpTime, nil
+}
+
+// prepareRetryTx prepares a new transaction for retrying by optionally updating the blockhash
+// and rebuilding the transaction with a bumped fee.
+func (txm *Txm) prepareRetryTx(ctx context.Context, baseTx solanaGo.Transaction, key string, msg *pendingTx, retryCount int, needBlockhashUpdate bool) (solanaGo.Transaction, error) {
+	// helper function to get latest blockhash
+	getLatestBlockhash := func(ctx context.Context) (*rpc.GetLatestBlockhashResult, error) {
+		client, err := txm.client.Get()
+		if err != nil {
+			return nil, fmt.Errorf("failed to get client: %w", err)
+		}
+		blockhashResult, err := client.LatestBlockhash(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get latest blockhash: %w", err)
+		}
+
+		return blockhashResult, nil
+	}
+
+	// We need to update blockhash in retryTx and pendingTx context if it has expired
+	if needBlockhashUpdate {
+		// Updates Blockhash and LastValidBlockHeight in the retry transaction
+		blockhashResult, err := getLatestBlockhash(ctx)
+		if err != nil {
+			return solanaGo.Transaction{}, fmt.Errorf("failed to get latest blockhash: %w", err)
+		}
+		baseTx.Message.RecentBlockhash = blockhashResult.Value.Blockhash
```

Comment:
I think this is ok since buildTx would take care of the signing below. The only thing like the previous comment is that we'd want to clear out `msg.signatures` too along with the entries for those in the signature map. But if we move the code to a TXM level loop, it might be easier to just use the `Remove` method to drop the expired tx from our storage and call `sendWithRetry` again on the baseTx. If we also move the `getLatestBlockhash` logic into `sendWithRetry` we wouldn't have to worry about that either